### ISSUE TITLE
Default add-ons bundled by Minishift should be installed automatically 

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#639]
At the moment a user has to explicitly install the default addons via `minishift addons install --defaults`. To be more user friendly we should install the bundled addons per default. This will require to keep some meta data somewhere in _$MINISHIFT_HOME_ which helps us to determine whether addons need to be installed. For example we need to distinguish the following cases:

- Fresh install (_$MINISHIFT_HOME_ does not exist)
- Existing _$MINISHIFT_HOME_ and user explicitly en- or disabled or in fact modified addons
- The Minishift binary got updated and there might be new versions of the addons

**Comment by jorgemoralespou** [comment]
+1
Also, as I mention in #637 this will make sense in that case.

**Comment by budhrg** [comment]
Just want to make sure that we are adding `admin-user` and `anyuid` as our default addons.
```
$ ls addons/
admin-user  anyuid  eap-templates  fuse-templates  registry-route
```
Do we need `registry-route` ?

**Comment by hferentschik** [comment]
admin-user and anyuid, yes
registry-route no (not for now at least)

**Comment by hferentschik** [comment]
@budhrg have you determined how you want to handle detection of whether or not to install default add-ons? Maybe let's document the behaviour as part of the issue?

**Comment by budhrg** [comment]
@hferentschik , I am just investigating on it right now. Looking for all aspects. Will document here with my findings and look for comments/suggestions.

**Comment by budhrg** [comment]
@hferentschik , I did the early investigation and found that during installation of defaults addons, the directories get created inside `$MINISHIFT_HOME/addons` with `$MINISHIFT_HOME/config/config.json`  as empty. `config.json` file get populated when we enable any addon.

So, I believe, we can have condition on whether the directories for default addons are created inside `$MINISHIFT_HOME/addons` or not to install default addons.
If there are no directories, then we install it which unpack the addon directories.
If there are already those addon directories, then we says "Default addons already installed."

Make sense?

**Comment by hferentschik** [comment]
> So, I believe, we can have condition on whether the directories for default addons are created inside $MINISHIFT_HOME/addons or not to install default addons.
If there are no directories, then we install it which unpack the addon directories.
If there are already those addon directories, then we says "Default addons already installed."

I think that falls a bit short. How does this cover the case of say add-ons installed and a new Minishift version being downloaded. Are we not going to handle this and if so how? The add-ons might have changed. We might have changed existing ones or added new ones. How do you handle this case?

And is it really enough to check the existence of the add-on directory. Should we not rather check for the existence of the default add-on itself. Say a user has used add-ons already, but only for a own custom addon. If he now gets a new version of Minishift, he won't get any default add-ons installed. 

One thing I was wondering about is, whether we should write a version string into say the config, something like currentMinisiftVersion. When we then execute a command we can match this version against the actual Minishift version and in this way detect that there is a new Minishift version. Now we have a chance to potentially check and update the add-ons.

**Comment by budhrg** [comment]
Thoughts on different cases : 
> Fresh install ($MINISHIFT_HOME does not exist)

For this one, we just unpack the default addons and should be trivial.

> Existing $MINISHIFT_HOME and user explicitly enabled or disabled or in fact modified addons

In this case, I am thinking of verifying each default addon for any change and then prompt user whether to be updated or not. like `Addon change detected. Want to reset 'anyuid' to default (y/n) ?`  or something along this line.

> The Minishift binary got updated and there might be new versions of the addons

In this case, if user change the default addon, the change get detected with new version. However, here we can also prompt user to reset to default which is based on the latest version of addons.

For detecting the change, I am thinking of having [`bytes.Equal`](https://golang.org/pkg/bytes/#Equal) of content of specific .addon file and the content after accessing from asset for specific .addon file.

I am also thinking that to get more control over default addons, we should restrict users for modifying it.

**Comment by hferentschik** [comment]
> I am also thinking that to get more control over default addons, we should restrict users for modifying it.

And how would you achieve this? We are not able to prevent this. The question is whether we just ignore modifications and for example update add-ons regardless or whether we try to handle it in some ways.

**Comment by hferentschik** [comment]
> Update command where PR is sent but review comments need to be implemented.

If we try to somehow utilise the 'update' command, how is this going to work? If the user explicitly does 'minishift update' and there is a new version, we download and install the new version and potentially do an add-on install/update. What if the user manually replaced the minishift binary. How are things working then? Do we expect updates to be done via the 'update' command?

**Comment by hferentschik** [comment]
I think we should do as part of this issue is the following. 

* As part of PersistentPreRun we check whether we have a new install/run by checking whether the MINISHIFT_HOME directory [exist|is empty] during PersistentPreRun. If it so, we will install the default add-ons. This will cover first time users or the case where a user completely empties the MINISHIFT_HOME directory.
* The second part will depend on pull request #813. Once it is merged we add a post update action which will be called after the binary is updated. We ask the user interactively whether default add-ons should be updated. If so, we unpack the latest add-ons regardless (the rub might be that we need to call the new executable via os.exec to get hold of the updated assets).

We skip having to deal with diffs and all. @budhrg wdyt?

**Comment by budhrg** [comment]
> We skip having to deal with diffs and all. @budhrg wdyt?

Seems fine to me :+1: 

> If so, we unpack the latest add-ons regardless (the rub might be that we need to call the new executable via os.exec to get hold of the updated assets).

We might need to handle similar thing in #813 . I am working on it for Windows side. See my last update here https://github.com/minishift/minishift/pull/813#issuecomment-301767331

**Comment by LalatenduMohanty** [comment]
For RPMs , if a installed RPM provide a config file and a new RPM is updated, we do the following

Either  we changes the config file to config.save and install the new one as config  or the new one is saved as config.new irrespective of whether the existing config file is changes or not.  Personally I feel the first option should work for us and it is simple to implement.  What I mean is if the file hs changed we should have a copy of this in the same dir as .save and install the new file.

**Comment by budhrg** [comment]
@hferentschik ,

> If so, we unpack the latest add-ons regardless (the rub might be that we need to call the new executable via os.exec to get hold of the updated assets).

I think even with getting the new executable via os.exec to get hold of updated assets, the code to unpack the addons is in the current binary execution context.
Assume following code: 
- code to fetch and update binary (old binary)
- get the new executable (new binary)
- code to unpack the addons (this comes in the execution context of old binary)

This is just my assumption.

**Comment by hferentschik** [comment]
Hmm I thought about ForkExec, but yes there is an issue. The alternative is to as part of the update, write a marker file into the MINISHIFT_HOME directory. In root.go we add some code which checks the marker. If it detects it, it does the requires post binary update actions and removes the marker.

**Comment by budhrg** [comment]
@hferentschik , how about executing `minishift addons install --defaults` once the binary is replaced?

**Comment by hferentschik** [comment]
Merged via pull request #990
